### PR TITLE
Allow the jitter entropy source to be used by the FIPS provider as an internal source

### PR DESCRIPTION
**Pull request by paulidale**
Doing so makes the FIPS provider non-compliant unless an entropy assessment and revalidation are also undertaken.

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: CHANGES.md
Comment (t8m) [inline CHANGES.md:35]:
```suggestion
   to undergo entropy source validation by the [CMVP], without this the FIPS provider
```

File: crypto/rand/build.info
Comment (t8m) [inline crypto/rand/build.info:21]: Nit: please drop this extra line

**Review by t8m — CHANGES_REQUESTED** [review]
Just two nits

File: CHANGES.md
Comment (paulidale) [reply CHANGES.md]: done.

File: crypto/rand/build.info
Comment (paulidale) [reply crypto/rand/build.info]: & done

**Comment by paulidale** [comment]
Comments addressed.

**Review by t8m — APPROVED** [review]
Approved, with the temporary commit dropped when merging

**Review by 0140454 — COMMENTED** [review]
There are confusions on viewing this PR, so I hope to get some guidance.

Please correct me if I am wrong. Thank you.

File: INSTALL.md
Comment (xnox) [inline INSTALL.md:854]: ... And re-validate on every change/submission due to chaining. If on the other hand it is unchained like in https://github.com/openssl/openssl/pull/25275 then an existing certificate could be reused. Standalone static library has been validated before https://csrc.nist.gov/projects/cryptographic-module-validation-program/entropy-validations/certificate/191 that could be marked for reuse, with public downloads available.

Although self-submission of jitterentropy-library is straight forward.

File: crypto/rand/rand_lib.c
Comment (xnox) [reply crypto/rand/rand_lib.c:606]: But even a public one, if it was certified, could be marked as propq "fips=yes".

File: crypto/rand/rand_lib.c
Comment (xnox) [reply crypto/rand/rand_lib.c:769]: Indeed, And it will safe on submissions too like mentioned in https://github.com/openssl/openssl/pull/25275

**Review by xnox — COMMENTED** [review]
Overall this is really nice, as it enables one to have just the fips provider using jitter entropy, rather than rely on top-level libcrypto. And it will enable to claim security strengths for generated keys.

The chaining is still problematic, it seems to add complexity for no gains; requiring additional submissions (as chain becomes part of conditioning, and creates need for ESV re-submissions all the time, rather than re-using existing ESV)

And anecdotally, unchained performance seems to be much better - but still need to create clean benchmarks for it. Even for non-fips case.

I will ping again about setting the ESV cert to mark it open for reuse. As then a certified prebuilt static library is available as a tarball download.

**Comment by xnox** [comment]
BTW 3.4 is not released yet, and if you only want to make jitter entropy source be available as an internal source, we can remove build options that make it a top level one.

File: crypto/rand/rand_lib.c
Comment (paulidale) [reply crypto/rand/rand_lib.c:606]: It really doesn't matter.  The fetch is from the internal library context which cannot be modified or altered by anything outside of the FIPS provider.  I'm just being careful and ignoring any global setting that might have been put in place, although I'm not sure how that would even happen.

File: crypto/rand/rand_lib.c
Comment (paulidale) [reply crypto/rand/rand_lib.c:769]: Inside the FIPS provider, #25415 replaces the primary DRBG with the health tests.  This means that, inside the FIPS provider, no DRBG is feeding another.

Outside the FIPS provider is pretty much irrelevant from NIST's perspective.  I kept the primary DRBG there because it greatly reduces the amount of entropy drawn from the system.

File: INSTALL.md
Comment (paulidale) [reply INSTALL.md:854]: As noted above, chaining is not happening inside the FIPS boundary.

**Comment by paulidale** [comment]
> BTW 3.4 is not released yet, and if you only want to make jitter entropy source be available as an internal source, we can remove build options that make it a top level one.

There is zero chance of this making 3.4 which is a bit annoying but wholly justifiable given its immaturity.  I certainly am not going to push for or even support its inclusion at this late stage.

We might be able to remove jitter as a public source with the justification that it will create ongoing legacy support.  However, I'm not all that convinced by that argument.  Jitter is useful outside of FIPS usage IMO.  One thing I did contemplate is making jitter the default source if it's been configured -- that seems sensible.  I doubt it would be accepted into 3.4 however.

File: INSTALL.md
Comment (paulidale) [reply INSTALL.md:854]: I don't think certificate reuse is possible when changing platforms.  The quality of the entropy is strongly tied to the CPU and it's environment.  Any change there is going to throw the computation off and require a re-assessment.

Some changes might be hideable from NIST (e.g. upgrading main memory or changing a disc drive), but the reality is everything will impact the jitter entropy quality to some extent.

File: crypto/rand/rand_lib.c
Comment (paulidale) [reply crypto/rand/rand_lib.c:736]: Will look into this tomorrow.  I suspect you are correct.  Too many negatives for my liking.

**Comment by paulidale** [comment]
> And it will enable to claim security strengths for generated keys.

I am planning on introducing an _entropy quality_ measure in a future PR.  We've no real concept of _strength_ in OpenSSL currently and I'd really like to have _strength_ made pervasive throughout of the entire library.

File: crypto/rand/rand_lib.c
Comment (0140454) [reply crypto/rand/rand_lib.c:606]: Got it. Thank you.

File: crypto/rand/rand_lib.c
Comment (0140454) [reply crypto/rand/rand_lib.c:769]: #25415 indeed replaces the primary DRBG with CRNGT via

```c
#ifdef FIPS_MODULE
    ret = rand_new_crngt(ctx, dgbl->seed);
#else
    ...
    ret = rand_new_drbg(ctx, dgbl->seed, PRIMARY_RESEED_INTERVAL,
                        PRIMARY_RESEED_TIME_INTERVAL, 1);
#endif
```

However, `#ifdef FIPS_MODULE` becomes `#if defined(FIPS_MODULE) && defined(OPENSSL_NO_FIPS_JITTER)` in this PR.

In the end, only FIPS provider built with `disable-fips-jitter` flag does this replacement.

Maybe we should keep `#ifdef FIPS_MODULE` here rather than change to `#if defined(FIPS_MODULE) ...`?

Comment (paulidale) [reply crypto/rand/rand_lib.c:769]: Good catch.  This isn't what I intended to do here.

**Comment by paulidale** [comment]
Last two commits are the new changes, I squashed the rest up for tidiness's sake.

I fixed the [issue raised above](https://github.com/openssl/openssl/pull/25498#discussion_r1769721936).

Fixed a locking problem with the CRNG test filter.

Removed the `use_df = 0` for the public and private DRBGs was just wrong and breaks FIPS compliance because the derivation function is mandatory under FIPS.  This was introduced in 3.2 by #16156.  The fix therefore doesn't need back porting since it isn't impacting any version undergoing FIPS validation (if 3.4 is to be validated, the fix needs to go there).

File: INSTALL.md
Comment (xnox) [reply INSTALL.md:854]: Please mark as resolved

**Comment by paulidale** [comment]
Dropping the temporary run the daily tests commit.

Ready for re-review.

**Comment by paulidale** [comment]
Thanks.

I've an open question about making jitter the default entropy source (outside of FIPS) if it is configured.
I'm also thinking about how to do entropy quality more nicely.

**Comment by paulidale** [comment]
Failing check addressed by #25582 and are not related to the changes here.

**Comment by paulidale** [comment]
Kicking CIs after #25582 has been merged.

**Comment by paulidale** [comment]
Still awaiting reviews @openssl/committers

**Comment by xnox** [comment]
I wonder how can I become `@openssl/committers` =)

**Comment by slontis** [comment]
> I wonder how can I become `@openssl/committers` =)

A good way it to keep up the good work you are doing contributing to openssl.
https://openssl-library.org/policies/general/committer-policy

The requirements to maintain committer status have recently been updated in the bylaws
https://openssl-library.org/about/bylaws/index.html

File: Configure
Comment (slontis) [inline Configure:966]: Is this right?
Could you potentially want to build fips provider with jitter but not use it in the default provider?

**Comment by paulidale** [comment]
Essentially, it boils down to submitting pull requests and doing reviews.
Do enough quality ones of each and you'll be noticed.

At least that's the current situation but with the OTC disappearing at some point the policy will change.  I expect the requirements of reviews and PRs won't change significantly.

File: Configure
Comment (paulidale) [reply Configure:966]: This just means that it is also available in the FIPS provider.  It won't be used unless config specifies it or the appropriate call is made to enable it.

I.e. this doesn't automatically make it _used_ in the default provider.

File: INSTALL.md
Comment (slontis) [inline INSTALL.md:852]:
```suggestion
3. have had the resulting FIPS provider certified by the [CMVP].
```

Comment (slontis) [reply INSTALL.md:852]: Never noticed the suggestion option until I just saw someone else using it..

File: NEWS.md
Comment (slontis) [inline NEWS.md:40]:
```suggestion
  * Allow the FIPS provider to optionally use the `JITTER` seed source.
```

File: README-FIPS.md
Comment (slontis) [inline README-FIPS.md:179]:
```suggestion
Note that that "enable-fips-jitter" option is only available for OpenSSL 3.5 or greater.
```

Comment (paulidale) [reply README-FIPS.md:179]: I'm not sure that this is required, this README will only be distributed in 3.5 and higher...

I guess it doesn't hurt.

Comment (paulidale) [reply README-FIPS.md:179]: I fixed the wording too :grin: 

> Note that the `enable-fips-jitter` option is only available in OpenSSL
versions 3.5 and later.

File: crypto/rand/rand_lib.c
Comment (slontis) [inline crypto/rand/rand_lib.c:638]: Adding a comment tag to the end on line #653 would help here :)

File: crypto/rand/rand_lib.c
Comment (slontis) [inline crypto/rand/rand_lib.c:737]:
```suggestion
#endif  /* FIPS_MODULE */
```

File: crypto/rand/rand_lib.c
Comment (slontis) [inline crypto/rand/rand_lib.c:771]: extra space before FIPS

File: Configure
Comment (slontis) [reply Configure:966]: Looking at build.info.. It does this

````
SUBDIRS=seeding

$RANDS_GOAL=../../libdefault.a ../../libfips.a

SOURCE[$RANDS_GOAL]=drbg.c test_rng.c drbg_ctr.c drbg_hash.c drbg_hmac.c
SOURCE[../../libdefault.a]=seed_src.c seed_src_jitter.c
SOURCE[../../libfips.a]=fips_crng_test.c
````

So the seed_src_jitter.c file is doing the check (which it probably shouldnt)..

The base and default provider both do this
````
static const OSSL_ALGORITHM deflt_rands[] = {
#ifndef OPENSSL_NO_JITTER
    { PROV_NAMES_JITTER, "provider=default", ossl_jitter_functions },
#endif
````

So enabling fips-jitter adds code to non fips source, which is what I disagree with.

**Comment by paulidale** [comment]
Feedback addressed.

File: doc/man7/EVP_RAND-JITTER.pod
Comment (slontis) [inline doc/man7/EVP_RAND-JITTER.pod:89]: Maybe just a NOTE would be better since the configuration options are not part of EVP_RAND-JITTER.pod

File: Configure
Comment (paulidale) [reply Configure:966]: Hmmmm.....  the other  jitter options also have to be specified to locate the jitter library and include directory and they don't have fips in the name.  Maybe I should document that fips-jitter also enables jitter or insist that both configure options are specified?  Having the code in the default provider is harmless since it won't be used but introduces a small amount of bloat.

Dissociating the two will be painful I suspect.

File: doc/man7/EVP_RAND-JITTER.pod
Comment (paulidale) [reply doc/man7/EVP_RAND-JITTER.pod]: Changed.

File: Configure
Comment (xnox) [reply Configure:966]: possibly a non-issue, as most do not perform a single configure and build of default and fips providers. It is quite typical to configure one tree and install headers, libcrypto, libssl from it; and another one to just get fips.so from. meaning most often default provider / libcrypto is discarded in the fips builds. But having this cleaner would be nice. Because if one only intents to use fips.so, one really does not need jitter in libcrypto.

Separately, I have been pondering a no-default build..... which doesn't link libdefault.a into libcrypto, such that default libcrypto only has base provider. Specifically for builds that only care about fips.so and nothing else. As it makes it possible to provide libcrypto for fips-certified-only environments witout ability to programatically load and use '-fips/fips=no' algorithms.

Comment (paulidale) [reply Configure:966]: Good point.  Once the first CVE fix comes out, split builds will be normal I suspect.

**Review by slontis — DISMISSED** [review]

**Comment by paulidale** [comment]
Rebase & squash with zero changes.

**Review by t8m — DISMISSED** [review]

**Review by hlandau — DISMISSED** [review]

File: CHANGES.md
Comment (xnox) [inline CHANGES.md:38]: ESV certification is separate from CMVP certification. When submitting for CMVP certification, one should already obtain ESV certificate. The same ESV certificate can be used multiple times for multiple CMVP submissions. The way the jitter entropy APIs are used, and the way it is linked statically, it is possible to validate jitter entropy library as a static library, and reuse it for multiple submissions. See for example public use document of https://csrc.nist.gov/projects/cryptographic-module-validation-program/entropy-validations/certificate/191

So this caveat is more strict than it actually is.

The idea is that many/most vendors have jitter entropy certified, and with these changes they can reuse their builds straight away. As most vendors have ESV for jitterentropy in kernel/openssl/gcrypt today, and this change makes it possible to reuse ESV, rather than require resubmissions each time. Previously chaining was causing the ESV to be non-reusable, and needed to reobtain ESV for every OpenSSL build change, due to OpenSSL DRBG being in the ESV boundary; which it no-longer is thanks to static linking of the standalone build/library and unchaining.

File: INSTALL.md
Comment (xnox) [inline INSTALL.md:854]: This is imprecise

```
Step 3 is Obtain ESV certificate for jitterentropy-library [ESV]
Step 4 have had the resulting FIPS provider certified by the [CMVP]

[ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations
```

Note that the same build/ESV can be used multiple times for different CMVP submissions of the cryptographic module, whilst compatible the same ESV cert can be reused over and over again.

File: NEWS.md
Comment (xnox) [inline NEWS.md:44]: `will need to be validated by the [ESV] before`

File: README-FIPS.md
Comment (xnox) [inline README-FIPS.md:178]: [ESV]

File: NEWS.md
Comment (xnox) [inline NEWS.md:2017]: replace [CMVP] with [ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

File: README-FIPS.md
Comment (xnox) [inline README-FIPS.md:184]: replace [CMVP] with [ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

File: doc/man7/EVP_RAND-JITTER.pod
Comment (xnox) [inline doc/man7/EVP_RAND-JITTER.pod:53]: [ESV] https://csrc.nist.gov/Projects/cryptographic-module-validation-program/entropy-validations

**Comment by xnox** [comment]
For avoidance of any ambiguity, whenever talking about entropy validation, please refer to the "ESV" - "Entropy Source Validation" as it is a separate validation and certification programme, related to CMVP, but distinct from it.

As one has to complete and receive ESV first, before one can complete a CMVP submission. But then later one can do multiple CMVP submissions with the same ESV.

Note as the changes are documentation only; this PR can be merged as is. And separately I can submit a PR to correct the docs / news / changes / comments. As there is not need to invalidate the current reviews & CI results.

**Comment by paulidale** [comment]
Amended references to CMVP to also mention ESV.

**Review by xnox — APPROVED** [review]
Also verified that only new changes pushed since previous approvals got dismissed are documentation only updates.

File: INSTALL.md
Comment (slontis) [inline INSTALL.md:852]:
```suggestion
3. obtain an [ESV] certificate for the |jitterentropy-library| and
```

**Review by slontis — APPROVED** [review]
Apart from a small NIT it looks good to me..

File: INSTALL.md
Comment (paulidale) [reply INSTALL.md:852]: Not sure we need the same link on sequential lines...

**Review by t8m — APPROVED** [review]

**Review by slontis — APPROVED** [review]

**Comment by paulidale** [comment]
Clean rebase and squash to recheck the CIs after recent merges.

**Comment by paulidale** [comment]
CI failures are a result of #25632 and aren't relevant to this PR.

**Pull request closed without merging**
